// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/marginfi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ComputeBudgetProgram, PublicKey, } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import BN from 'bn.js';
import { SEEDS } from '../constants';
import { MarginfiAccount, MarginfiClient, PDA_BANK_LIQUIDITY_VAULT_AUTH_SEED, PDA_BANK_LIQUIDITY_VAULT_SEED, getConfig, } from '@mercurial-finance/marginfi-client-v2';
import { getOrCreateATAInstruction } from '../utils';
export default class MarginFiHandler {
    constructor(program) {
        this.connection = program.provider.connection;
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const marginfiClient = yield MarginfiClient.fetch(getConfig(), {}, this.connection);
            const strategyBuffer = strategy.pubkey.toBuffer();
            const [marginfiPda] = PublicKey.findProgramAddressSync([Buffer.from(SEEDS.MARGINFI_ACCOUNT), strategyBuffer], program.programId);
            const marginfiAccount = yield MarginfiAccount.fetch(marginfiPda, marginfiClient);
            const group = marginfiAccount.group;
            const bank = group.getBankByMint(vaultState.tokenMint);
            if (!bank)
                throw new Error('No bank found');
            const [collateralVault] = PublicKey.findProgramAddressSync([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const [strategyOwner] = PublicKey.findProgramAddressSync([Buffer.from(SEEDS.MARGINFI_STRATEGY), strategyBuffer], program.programId);
            const [tokenAccount, createTokenAccountIx] = yield getOrCreateATAInstruction(vaultState.tokenMint, strategyOwner, this.connection, {
                payer: walletPubKey,
            });
            createTokenAccountIx && preInstructions.push(createTokenAccountIx);
            const strategyReserveBuffer = strategy.state.reserve.toBuffer();
            const [bankLiquidityVault] = PublicKey.findProgramAddressSync([PDA_BANK_LIQUIDITY_VAULT_SEED, strategyReserveBuffer], marginfiClient.programId);
            const [bankLiquidityVaultAuth] = PublicKey.findProgramAddressSync([PDA_BANK_LIQUIDITY_VAULT_AUTH_SEED, strategyReserveBuffer], marginfiClient.programId);
            const observationAccounts = marginfiAccount.getHealthCheckAccounts([bank]);
            const accounts = [
                { pubkey: strategyOwner, isWritable: true },
                { pubkey: bank.group, isWritable: true },
                { pubkey: marginfiAccount.publicKey, isWritable: true },
                { pubkey: tokenAccount, isWritable: true },
                { pubkey: bankLiquidityVault, isWritable: true },
                { pubkey: bankLiquidityVaultAuth, isWritable: true },
                ...observationAccounts,
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: account.isWritable,
                    isSigner: false,
                });
            }
            // Do not remove: to resolve limit of computation and log
            const additionalComputeBudgetInstruction = ComputeBudgetProgram.setComputeUnitLimit({
                units: 800000,
            });
            preInstructions.push(additionalComputeBudgetInstruction);
            const txAccounts = {
                vault,
                strategy: new PublicKey(strategy.pubkey),
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: marginfiClient.programId,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
