// @switchboard-xyz/solana.js/lib/generated/attestation-program/instructions/functionRequestVerify.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([
    types.FunctionRequestVerifyParams.layout("params"),
]);
export function functionRequestVerify(program, args, accounts, programId = program.attestationProgramId) {
    const keys = [
        { pubkey: accounts.request, isSigner: false, isWritable: true },
        {
            pubkey: accounts.functionEnclaveSigner,
            isSigner: true,
            isWritable: false,
        },
        { pubkey: accounts.escrow, isSigner: false, isWritable: true },
        { pubkey: accounts.function, isSigner: false, isWritable: true },
        { pubkey: accounts.functionEscrow, isSigner: false, isWritable: true },
        { pubkey: accounts.verifierQuote, isSigner: false, isWritable: false },
        {
            pubkey: accounts.verifierEnclaveSigner,
            isSigner: true,
            isWritable: false,
        },
        { pubkey: accounts.verifierPermission, isSigner: false, isWritable: false },
        { pubkey: accounts.state, isSigner: false, isWritable: false },
        { pubkey: accounts.attestationQueue, isSigner: false, isWritable: false },
        { pubkey: accounts.receiver, isSigner: false, isWritable: true },
        { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    ];
    const identifier = Buffer.from([179, 6, 88, 97, 232, 112, 143, 253]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.FunctionRequestVerifyParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
