// @switchboard-xyz/solana.js/lib/generated/attestation-program/instructions/functionVerify.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([
    types.FunctionVerifyParams.layout("params"),
]);
export function functionVerify(program, args, accounts, programId = program.attestationProgramId) {
    const keys = [
        { pubkey: accounts.function, isSigner: false, isWritable: true },
        {
            pubkey: accounts.functionEnclaveSigner,
            isSigner: true,
            isWritable: false,
        },
        { pubkey: accounts.verifier, isSigner: false, isWritable: false },
        { pubkey: accounts.verifierSigner, isSigner: true, isWritable: false },
        { pubkey: accounts.verifierPermission, isSigner: false, isWritable: false },
        { pubkey: accounts.escrowWallet, isSigner: false, isWritable: false },
        { pubkey: accounts.escrowTokenWallet, isSigner: false, isWritable: true },
        { pubkey: accounts.receiver, isSigner: false, isWritable: true },
        { pubkey: accounts.attestationQueue, isSigner: false, isWritable: false },
        { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    ];
    const identifier = Buffer.from([210, 108, 154, 138, 198, 14, 53, 191]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.FunctionVerifyParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
