// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/WalletInitParams.js
import * as borsh from "@coral-xyz/borsh";
export class WalletInitParams {
    name;
    maxLen;
    constructor(fields) {
        this.name = fields.name;
        this.maxLen = fields.maxLen;
    }
    static layout(property) {
        return borsh.struct([borsh.vecU8("name"), borsh.option(borsh.u32(), "maxLen")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new WalletInitParams({
            name: new Uint8Array(obj.name.buffer, obj.name.byteOffset, obj.name.length),
            maxLen: obj.maxLen,
        });
    }
    static toEncodable(fields) {
        return {
            name: Buffer.from(fields.name.buffer, fields.name.byteOffset, fields.name.length),
            maxLen: fields.maxLen,
        };
    }
    toJSON() {
        return {
            name: Array.from(this.name.values()),
            maxLen: this.maxLen,
        };
    }
    static fromJSON(obj) {
        return new WalletInitParams({
            name: Uint8Array.from(obj.name),
            maxLen: obj.maxLen,
        });
    }
    toEncodable() {
        return WalletInitParams.toEncodable(this);
    }
}
