// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/WalletInitParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletInitParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class WalletInitParams {
    name;
    maxLen;
    constructor(fields) {
        this.name = fields.name;
        this.maxLen = fields.maxLen;
    }
    static layout(property) {
        return borsh.struct([borsh.vecU8("name"), borsh.option(borsh.u32(), "maxLen")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new WalletInitParams({
            name: new Uint8Array(obj.name.buffer, obj.name.byteOffset, obj.name.length),
            maxLen: obj.maxLen,
        });
    }
    static toEncodable(fields) {
        return {
            name: Buffer.from(fields.name.buffer, fields.name.byteOffset, fields.name.length),
            maxLen: fields.maxLen,
        };
    }
    toJSON() {
        return {
            name: Array.from(this.name.values()),
            maxLen: this.maxLen,
        };
    }
    static fromJSON(obj) {
        return new WalletInitParams({
            name: Uint8Array.from(obj.name),
            maxLen: obj.maxLen,
        });
    }
    toEncodable() {
        return WalletInitParams.toEncodable(this);
    }
}
exports.WalletInitParams = WalletInitParams;
