// @mercurial-finance/dynamic-amm-sdk/dist/esm/src/amm/tests/utils/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Wallet } from '@project-serum/anchor';
import { TOKEN_PROGRAM_ID, Token } from '@solana/spl-token';
import { Keypair, LAMPORTS_PER_SOL } from '@solana/web3.js';
export const airDropSol = (connection, publicKey, amount = 1) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const airdropSignature = yield connection.requestAirdrop(publicKey, amount * LAMPORTS_PER_SOL);
        const latestBlockHash = yield connection.getLatestBlockhash();
        yield connection.confirmTransaction({
            blockhash: latestBlockHash.blockhash,
            lastValidBlockHeight: latestBlockHash.lastValidBlockHeight,
            signature: airdropSignature,
        }, connection.commitment);
    }
    catch (error) {
        console.error(error);
        throw error;
    }
});
export const airDropSolIfBalanceNotEnough = (connection, publicKey, balance = 1) => __awaiter(void 0, void 0, void 0, function* () {
    const walletBalance = yield connection.getBalance(publicKey);
    if (walletBalance < balance * LAMPORTS_PER_SOL) {
        yield airDropSol(connection, publicKey);
    }
});
export const getOrCreateATA = (connection, mint, owner, payer) => __awaiter(void 0, void 0, void 0, function* () {
    const token = new Token(connection, mint, TOKEN_PROGRAM_ID, payer);
    const ata = yield token.getOrCreateAssociatedAccountInfo(owner);
    return ata.address;
});
export const mockWallet = new Wallet(Keypair.generate());
// export const MAINNET = {
//   connection: new Connection(process.env.MAINNET_RPC_ENDPOINT as string),
//   cluster: 'mainnet-beta',
// };
// export const DEVNET = {
//   connection: new Connection('https://api.devnet.solana.com/', {
//     commitment: 'confirmed',
//   }),
//   cluster: 'devnet',
// };
